BibCheck: geocode all the things

jacquerie opened the pull request:
As for #210, I haven't run or tested this code, but it looks right, so it must also be correct : )

* First commit fixes my twitchy eye when I look at that previous commit.
* Second commit geocodes all Conferences and Institutions that could have it. It's possible that we run into the API limits on this one; it should be enough to restart the task the next day (it looks idempotent).

kaplun commented on bibcheck/plugins/geocode_all_the_things.py:
```diff
@@ -0,0 +1,83 @@
+def geocode(record):
+    """Uses the Google Maps API to geocode a record."""
+    def _get_approximate_address(record):
+        if len(list(record.iterfield('111__c'))):
```
Why not simply 
```python
if record_get_field_values(record, '111', code='c'):
```
?

kaplun replied:
Wait: 111__c? has no results.
https://inspirehep.net/search?ln=en&cc=Institutions&ln=en&cc=Institutions&p=111__c%3A**&action_search=Search&sf=institutionname&so=a&rm=&rg=25&sc=0&of=hb

kaplun replied:
OK. Can you comment that this is the case of conferences?

kaplun commented on bibcheck/plugins/geocode_all_the_things.py:
```diff
@@ -0,0 +1,83 @@
+    def _has_no_latitude_or_longitude(record):
+        return (
+            len(list(record.iterfield('034__f'))) == 0 and
+            len(list(record.iterfield('034__d'))) == 0)
```
What about even `or`, so that even in case of partial coordinate, we just redo them from scratch.

kaplun commented on bibcheck/plugins/geocode_all_the_things.py:
```diff
@@ -0,0 +1,83 @@
+
+    approximate_address = _get_approximate_address(record)
+    while approximate_address:
+        response = GMAPS.geocode(', '.join(approximate_address))
```
After we reach the Google API's limit, what will happen? raised exception? I need to think how to do properly stop the process.

jacquerie replied:
Yes: https://github.com/googlemaps/google-maps-services-python/blob/2ccf0b2912019341aa60aeb65fc36ca6d9d02a56/googlemaps/client.py#L258-L259

Is that a problem? I'm not familiar with how BibCheck deals with exceptions.

kaplun replied:
Probably not. As long as the script know it has to crash and has hence to reconsider all the records and not just the modified ones.

jacquerie commented on bibcheck/plugins/geocode_all_the_things.py:
```diff
@@ -0,0 +1,101 @@
+        record.add_field('034__', '', subfields=subfields)
+
+    def _remove_all_latitudes_and_longitudes(record):
+        for pos, _ in reversed(record.iterfield('034__f')):
```
Is this deleting ALL `034__f`? At least these two cases:
```xml
<record>
  <datafield tag="034" ind1=" " ind2=" ">
    <subfield code="d">foo</subfield>
    <subfield code="f">bar</subfield>
  </datafield>
  <datafield tag="034" ind1=" " ind2=" ">
    <subfield code="d">baz</subfield>
  </datafield>
</record>
```
and
```xml
<record>
  <datafield tag="034" ind1=" " ind2=" ">
    <subfield code="d">foo</subfield>
  </datafield>
  <datafield tag="034" ind1=" " ind2=" ">
    <subfield code="d">bar</subfield>
    <subfield code="f">baz</subfield>
  </datafield>
</record>
```

tsgit commented on bibcheck/plugins/geocode_all_the_things.py:
```diff
@@ -0,0 +1,101 @@
+        subfields = [('f', str(latitude)), ('d', str(longitude))]
+        record.add_field('034__', '', subfields=subfields)
+
+    def _remove_all_latitudes_and_longitudes(record):
```
question: is there other content in MARC tag 034 for Institutions?
if not, why not `del record["034"]`
and if so, you could combine 2 loop into one with  
`... reversed(record.iterfields(['034__f', '034__d']))`

jacquerie commented:
What's blocking this? I would really like to have latitudes and longitudes in the next dump...

kaplun commented:
IMHO it's OK to go. @tsgit could you take care of merge+deploy? This bibcheck should not run in the regular run because it will crash by definition after 3000 requests (due to reaching Google API limit). So it should be scheduled in a request that is run every 24hours.

tsgit commented on bibcheck/plugins/geocode_all_the_things.py:
```diff
@@ -0,0 +1,97 @@
+        record.add_field('034__', '', subfields=subfields)
+
+    def _remove_all_latitudes_and_longitudes(record):
+        del record['034__']
```
the record dict is keyed by MARC tag without subfield, so I think this must be `del record['034']`

```
In [1]: from invenio.bibcheck_task import AmendableRecord

In [2]: from invenio.bibedit_utils import get_bibrecord

In [3]: rec = AmendableRecord(get_bibrecord(902626))

In [4]: rec['034']
[([('d', '10.202053'), ('f', '56.1632189')], ' ', ' ', '', 22)]
```

tsgit replied:
also, naturally you can only delete the key if it is present
```
2016-10-19 00:18:49 --> Traceback is:
2016-10-19 00:18:49 --> * 2016-10-19 00:18:49 -> KeyError: '034' (geocode_all_the_things.py:90:_remove_all_latitudes_and_longitudes)
2016-10-19 00:18:49 --> ** User details
2016-10-19 00:18:49 --> No client information available
2016-10-19 00:18:49 --> ** Traceback details 
2016-10-19 00:18:49 --> Traceback (most recent call last):
2016-10-19 00:18:49 -->   File "/usr/lib64/python2.6/site-packages/invenio/bibtask.py", line 620, in task_init
2016-10-19 00:18:49 -->     ret = _task_run(task_run_fnc)
2016-10-19 00:18:49 -->   File "/usr/lib64/python2.6/site-packages/invenio/bibtask.py", line 1184, in _task_run
2016-10-19 00:18:49 -->     if callable(task_run_fnc) and task_run_fnc():
2016-10-19 00:18:49 -->   File "/usr/lib64/python2.6/site-packages/invenio/bibcheck_task.py", line 613, in task_run_core
2016-10-19 00:18:49 -->     check_record(rule, record)
2016-10-19 00:18:49 -->   File "/usr/lib64/python2.6/site-packages/invenio/bibcheck_task.py", line 691, in check_record
2016-10-19 00:18:49 -->     return plugin["check_record"](record, **rule["checker_params"])
2016-10-19 00:18:49 -->   File "/opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py", line 93, in check_record
2016-10-19 00:18:49 -->     _remove_all_latitudes_and_longitudes(record)
2016-10-19 00:18:49 -->   File "/opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py", line 90, in _remove_all_latitudes_and_longitudes
2016-10-19 00:18:49 -->     del record['034']
2016-10-19 00:18:49 --> KeyError: '034'
2016-10-19 00:18:49 --> ** Stack frame details
2016-10-19 00:18:49 --> Frame _remove_all_latitudes_and_longitudes in /opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py at line 90
2016-10-19 00:18:49 --> -------------------------------------------------------------------------------
2016-10-19 00:18:49 -->         87         record.add_field('034__', '', subfields=subfields)
2016-10-19 00:18:49 -->         88 
2016-10-19 00:18:50 -->         89     def _remove_all_latitudes_and_longitudes(record):
2016-10-19 00:18:50 --> ---->   90         del record['034']
2016-10-19 00:18:50 -->         91 
2016-10-19 00:18:50 -->         92     if _has_no_latitude_or_longitude(record) and _should_have_them(record):
2016-10-19 00:18:50 -->         93         _remove_all_latitudes_and_longitudes(record)
2016-10-19 00:18:50 --> -------------------------------------------------------------------------------
```

tsgit commented:
there's 11k Institutions records so after 4 days of batches of 3000 we will be caught up and then it can be a regular periodic task

tsgit commented:
another problem:
```
2016-10-19 00:25:16 --> Task #904223 finished. [RUNNING]
2016-10-19 00:25:16 --> Unexpected error occurred: global name 'a_values' is not defined.
2016-10-19 00:25:16 --> Traceback is:
2016-10-19 00:25:16 --> * 2016-10-19 00:25:16 -> NameError: global name 'a_values' is not defined (geocode_all_the_things.py:56:_get_approximate_address)
2016-10-19 00:25:16 --> ** User details
2016-10-19 00:25:16 --> No client information available
2016-10-19 00:25:16 --> ** Traceback details 
2016-10-19 00:25:16 --> Traceback (most recent call last):
2016-10-19 00:25:16 -->   File "/usr/lib64/python2.6/site-packages/invenio/bibtask.py", line 620, in task_init
2016-10-19 00:25:16 -->     ret = _task_run(task_run_fnc)
2016-10-19 00:25:16 -->   File "/usr/lib64/python2.6/site-packages/invenio/bibtask.py", line 1184, in _task_run
2016-10-19 00:25:16 -->     if callable(task_run_fnc) and task_run_fnc():
2016-10-19 00:25:16 -->   File "/usr/lib64/python2.6/site-packages/invenio/bibcheck_task.py", line 613, in task_run_core
2016-10-19 00:25:16 -->     check_record(rule, record)
2016-10-19 00:25:16 -->   File "/usr/lib64/python2.6/site-packages/invenio/bibcheck_task.py", line 691, in check_record
2016-10-19 00:25:16 -->     return plugin["check_record"](record, **rule["checker_params"])
2016-10-19 00:25:16 -->   File "/opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py", line 96, in check_record
2016-10-19 00:25:16 -->     latitude, longitude = geocode(record)
2016-10-19 00:25:16 -->   File "/opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py", line 62, in geocode
2016-10-19 00:25:16 -->     approximate_address = _get_approximate_address(record)
2016-10-19 00:25:16 -->   File "/opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py", line 56, in _get_approximate_address
2016-10-19 00:25:16 -->     result = [el for el in a_values if zipcode not in el]
2016-10-19 00:25:16 --> NameError: global name 'a_values' is not defined
2016-10-19 00:25:16 --> ** Stack frame details
2016-10-19 00:25:16 --> Frame _get_approximate_address in /opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py at line 56
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->         53             country = record_get_field_value(record, '371', code='d')
2016-10-19 00:25:16 -->         54             addresses = record_get_field_values(record, '371', code='a')
2016-10-19 00:25:16 -->         55 
2016-10-19 00:25:16 --> ---->   56             result = [el for el in a_values if zipcode not in el]
2016-10-19 00:25:16 -->         57             result.extend([city, zipcode, country])
2016-10-19 00:25:16 -->         58             result = [el for el in addresses if el]
2016-10-19 00:25:16 -->         59 
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->                         city =  "'Aquila'"
2016-10-19 00:25:16 -->                    addresses =  "['Via Vetoio Localit\\xc3\\xa0 Coppito', 'I-67100 Aquila']"
2016-10-19 00:25:16 -->             _is_a_conference =  '<function _is_a_conference at 0x1b2ce500>'
2016-10-19 00:25:16 -->                      country =  "'Italy'"
2016-10-19 00:25:16 -->                      zipcode =  "'67100'"
2016-10-19 00:25:16 -->                         _[1] =  '[]'
2016-10-19 00:25:16 -->                       record =  '{\'856\': [([(\'u\', \'http://www.fisica.aquila.infn.it/\')], \'4\', \' \', \'\', 3)], \'961\': [([(\'c\', \'2008-12-03\')], \' \', \' \', \'\', 5), ([(\'x\', \'1989-08-08\')], \' \', \' \', \'\', 6)], \'970\': [([(\'a\', \'INST-248\')], \' \', \' \', \'\', 7)], \'980\': [([(\'b\', \'HEP200\'), (\'b\', \'PPF\'), (\'b\', \'WEB\')], \' \', \' \', \'\', 8), ([(\'a\', \'INSTITUTION\')], \' \', \' \', \'\', 9)], \'909\': [([(\'o\', \'oai:inspirehep.net:902644\'), (\'p\', \'INSPIRE:Institutions\')],  [...]
2016-10-19 00:25:16 --> Frame geocode in /opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py at line 62
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->         59 
2016-10-19 00:25:16 -->         60             return result
2016-10-19 00:25:16 -->         61 
2016-10-19 00:25:16 --> ---->   62     approximate_address = _get_approximate_address(record)
2016-10-19 00:25:16 -->         63     while approximate_address:
2016-10-19 00:25:16 -->         64         response = GMAPS.geocode(', '.join(approximate_address))
2016-10-19 00:25:16 -->         65         if response:
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->                       record =  '{\'856\': [([(\'u\', \'http://www.fisica.aquila.infn.it/\')], \'4\', \' \', \'\', 3)], \'961\': [([(\'c\', \'2008-12-03\')], \' \', \' \', \'\', 5), ([(\'x\', \'1989-08-08\')], \' \', \' \', \'\', 6)], \'970\': [([(\'a\', \'INST-248\')], \' \', \' \', \'\', 7)], \'980\': [([(\'b\', \'HEP200\'), (\'b\', \'PPF\'), (\'b\', \'WEB\')], \' \', \' \', \'\', 8), ([(\'a\', \'INSTITUTION\')], \' \', \' \', \'\', 9)], \'909\': [([(\'o\', \'oai:inspirehep.net:902644\'), (\'p\', \'INSPIRE:Institutions\')],  [...]
2016-10-19 00:25:16 -->         _get_approximate_address =  '<function _get_approximate_address at 0x1b2ce578>'
2016-10-19 00:25:16 --> Frame check_record in /opt/cds-invenio/lib/python/invenio/bibcheck_plugins/geocode_all_the_things.py at line 96
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->         93     if _has_no_latitude_or_longitude(record) and _should_have_them(record):
2016-10-19 00:25:16 -->         94         _remove_all_latitudes_and_longitudes(record)
2016-10-19 00:25:16 -->         95 
2016-10-19 00:25:16 --> ---->   96         latitude, longitude = geocode(record)
2016-10-19 00:25:16 -->         97         if latitude and longitude:
2016-10-19 00:25:16 -->         98             _add_latitude_and_longitude(record, latitude, longitude)
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->         _remove_all_latitudes_and_longitudes =  '<function _remove_all_latitudes_and_longitudes at 0x17b41320>'
2016-10-19 00:25:16 -->         _add_latitude_and_longitude =  '<function _add_latitude_and_longitude at 0x1b23eed8>'
2016-10-19 00:25:16 -->                       record =  '{\'856\': [([(\'u\', \'http://www.fisica.aquila.infn.it/\')], \'4\', \' \', \'\', 3)], \'961\': [([(\'c\', \'2008-12-03\')], \' \', \' \', \'\', 5), ([(\'x\', \'1989-08-08\')], \' \', \' \', \'\', 6)], \'970\': [([(\'a\', \'INST-248\')], \' \', \' \', \'\', 7)], \'980\': [([(\'b\', \'HEP200\'), (\'b\', \'PPF\'), (\'b\', \'WEB\')], \' \', \' \', \'\', 8), ([(\'a\', \'INSTITUTION\')], \' \', \' \', \'\', 9)], \'909\': [([(\'o\', \'oai:inspirehep.net:902644\'), (\'p\', \'INSPIRE:Institutions\')],  [...]
2016-10-19 00:25:16 -->            _should_have_them =  '<function _should_have_them at 0x1b23ecf8>'
2016-10-19 00:25:16 -->         _has_no_latitude_or_longitude =  '<function _has_no_latitude_or_longitude at 0x1b23e050>'
2016-10-19 00:25:16 --> Frame check_record in /usr/lib64/python2.6/site-packages/invenio/bibcheck_task.py at line 691
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->        688     record.set_rule(rule)
2016-10-19 00:25:16 -->        689     plugin = plugins[rule["check"]]
2016-10-19 00:25:16 -->        690     if not record.is_dummy():
2016-10-19 00:25:16 --> ---->  691         return plugin["check_record"](record, **rule["checker_params"])
2016-10-19 00:25:16 -->        692 
2016-10-19 00:25:16 -->        693 def check_records(rule, records):
2016-10-19 00:25:16 -->        694     """
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->                       record =  '{\'856\': [([(\'u\', \'http://www.fisica.aquila.infn.it/\')], \'4\', \' \', \'\', 3)], \'961\': [([(\'c\', \'2008-12-03\')], \' \', \' \', \'\', 5), ([(\'x\', \'1989-08-08\')], \' \', \' \', \'\', 6)], \'970\': [([(\'a\', \'INST-248\')], \' \', \' \', \'\', 7)], \'980\': [([(\'b\', \'HEP200\'), (\'b\', \'PPF\'), (\'b\', \'WEB\')], \' \', \' \', \'\', 8), ([(\'a\', \'INSTITUTION\')], \' \', \' \', \'\', 9)], \'909\': [([(\'o\', \'oai:inspirehep.net:902644\'), (\'p\', \'INSPIRE:Institutions\')],  [...]
2016-10-19 00:25:16 -->                      plugins =  "{'crossref_checker': {'name': 'crossref_checker', 'check_record': None, 'mandatory_args': set([]), 'batch': True, 'check_records': <function check_records at 0x1b2b9500>, 'all_args': set(['doi_field'])}, 'mandatory': {'name': 'mandatory', 'check_record': <function check_record at 0x1b23eaa0>, 'mandatory_args': set(['fields']), 'batch': False, 'check_records': None, 'all_args': set(['fields'])}, 'regexp_replace': {'name': 'regexp_replace', 'check_record': <function check_record at 0x1b246938>, ' [...]
2016-10-19 00:25:16 -->                         rule =  "{'holdingpen': False, 'filter_collection': 'Institutions', 'checker_params': {}, 'name': 'geocode', 'check': 'geocode_all_the_things'}"
2016-10-19 00:25:16 -->                       plugin =  "{'name': 'geocode_all_the_things', 'check_record': <function check_record at 0x1b1d5d70>, 'mandatory_args': set([]), 'batch': False, 'check_records': None, 'all_args': set([])}"
2016-10-19 00:25:16 --> Frame task_run_core in /usr/lib64/python2.6/site-packages/invenio/bibcheck_task.py at line 613
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->        610                 rule_recids = recids_for_rules[rule_name]
2016-10-19 00:25:16 -->        611                 task_sleep_now_if_required(can_stop_too=True)
2016-10-19 00:25:16 -->        612                 if record_id in rule_recids:
2016-10-19 00:25:16 --> ---->  613                     check_record(rule, record)
2016-10-19 00:25:16 -->        614 
2016-10-19 00:25:16 -->        615             if record.amended:
2016-10-19 00:25:16 -->        616                 if record.holdingpen:
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->             recids_for_rules =  "{'geocode': intbitset([902644])}"
2016-10-19 00:25:16 -->              update_database =  'False'
2016-10-19 00:25:16 -->                   all_recids =  'intbitset([902644])'
2016-10-19 00:25:16 -->                    rule_name =  "'geocode'"
2016-10-19 00:25:16 -->                            i =  '0'
2016-10-19 00:25:16 -->         records_to_upload_replace =  '[]'
2016-10-19 00:25:16 -->               rules_to_reset =  "set(['geocode'])"
2016-10-19 00:25:16 -->                        batch =  '[(0, 902644, {\'856\': [([(\'u\', \'http://www.fisica.aquila.infn.it/\')], \'4\', \' \', \'\', 3)], \'961\': [([(\'c\', \'2008-12-03\')], \' \', \' \', \'\', 5), ([(\'x\', \'1989-08-08\')], \' \', \' \', \'\', 6)], \'970\': [([(\'a\', \'INST-248\')], \' \', \' \', \'\', 7)], \'980\': [([(\'b\', \'HEP200\'), (\'b\', \'PPF\'), (\'b\', \'WEB\')], \' \', \' \', \'\', 8), ([(\'a\', \'INSTITUTION\')], \' \', \' \', \'\', 9)], \'909\': [([(\'o\', \'oai:inspirehep.net:902644\'), (\'p\', \'INSPIRE:Insti [...]
2016-10-19 00:25:16 -->         records_to_submit_tickets =  '[]'
2016-10-19 00:25:16 -->             progress_percent =  '0'
2016-10-19 00:25:16 -->                 single_rules =  "set(['geocode'])"
2016-10-19 00:25:16 -->                       record =  '{\'856\': [([(\'u\', \'http://www.fisica.aquila.infn.it/\')], \'4\', \' \', \'\', 3)], \'961\': [([(\'c\', \'2008-12-03\')], \' \', \' \', \'\', 5), ([(\'x\', \'1989-08-08\')], \' \', \' \', \'\', 6)], \'970\': [([(\'a\', \'INST-248\')], \' \', \' \', \'\', 7)], \'980\': [([(\'b\', \'HEP200\'), (\'b\', \'PPF\'), (\'b\', \'WEB\')], \' \', \' \', \'\', 8), ([(\'a\', \'INSTITUTION\')], \' \', \' \', \'\', 9)], \'909\': [([(\'o\', \'oai:inspirehep.net:902644\'), (\'p\', \'INSPIRE:Institutions\')],  [...]
2016-10-19 00:25:16 -->         records_to_upload_holdingpen =  '[]'
2016-10-19 00:25:16 -->                        rules =  "{'geocode': {'holdingpen': False, 'filter_collection': 'Institutions', 'checker_params': {}, 'name': 'geocode', 'check': 'geocode_all_the_things'}}"
2016-10-19 00:25:16 -->                         rule =  "{'holdingpen': False, 'filter_collection': 'Institutions', 'checker_params': {}, 'name': 'geocode', 'check': 'geocode_all_the_things'}"
2016-10-19 00:25:16 -->                  rule_recids =  'intbitset([902644])'
2016-10-19 00:25:16 -->                      plugins =  "{'crossref_checker': {'name': 'crossref_checker', 'check_record': None, 'mandatory_args': set([]), 'batch': True, 'check_records': <function check_records at 0x1b2b9500>, 'all_args': set(['doi_field'])}, 'mandatory': {'name': 'mandatory', 'check_record': <function check_record at 0x1b23eaa0>, 'mandatory_args': set(['fields']), 'batch': False, 'check_records': None, 'all_args': set(['fields'])}, 'regexp_replace': {'name': 'regexp_replace', 'check_record': <function check_record at 0x1b246938>, ' [...]
2016-10-19 00:25:16 -->                    record_id =  '902644'
2016-10-19 00:25:16 -->                  batch_rules =  'set([])'
2016-10-19 00:25:16 --> Frame _task_run in /usr/lib64/python2.6/site-packages/invenio/bibtask.py at line 1184
2016-10-19 00:25:16 --> -------------------------------------------------------------------------------
2016-10-19 00:25:16 -->       1181 
2016-10-19 00:25:16 -->       1182     sleeptime = _TASK_PARAMS['sleeptime']
2016-10-19 00:25:16 -->       1183     try:
2016-10-19 00:25:16 --> ----> 1184         if callable(task_run_fnc) and task_run_fnc():
2016-10-19 00:25:16 -->       1185             task_update_status("DONE")
2016-10-19 00:25:16 -->       1186         else:
2016-10-19 00:25:16 -->       1187             task_update_status("DONE WITH ERRORS")
```
```

tsgit commented:
pylint also says that `os` is unused

```
$ pylint geocode_all_the_things.py 
No config file found, using default configuration
************* Module geocode_all_the_things
E: 56,35: Undefined variable 'a_values' (undefined-variable)
C: 74, 0: Missing function docstring (missing-docstring)
C: 89, 4: Invalid function name "_remove_all_latitudes_and_longitudes" (invalid-name)
W: 30, 0: Unused import os (unused-import)

...
Your code has been rated at 8.10/10 (previous run: 8.10/10, +0.00)
```

tsgit commented:
for https://inspirehep.net/record/902644 I got

https://www.google.com/maps/place/42%C2%B024'48.4%22N+13%C2%B029'50.7%22E/@42.413435,13.4952373,17z/data=!4m5!3m4!1s0x0:0x0!8m2!3d42.413435!4d13.497426

which is close to L'Aquila , but seems a bit on countryside. Is that expected with a good address as 
```
 Dipartimento di Fisica
Via Vetoio Località Coppito
I-67100 Aquila
Italy 
```

jacquerie commented:
I cannot answer the review comments there, so I'll write my answers here:

* Removed `import os`.
* Whoops! Fixed the `a_values` thing. I changed my name a few times on how to call the variables and left them in an inconsistent state.
* Yeah, that geocode is super precise yet wrong. That's why I wanted to only use the city when geocoding, so that we don't see these kind of errors. TBH, we really just want some approximate values there, so that we can put them on a map at a very low zoom level and they kind of look right. If you load your GPS from INSPIRE you're not using it right...

kaplun commented:
> Yeah, that geocode is super precise yet wrong. That's why I wanted to only use the city when geocoding, so that we don't see these kind of errors. TBH, we really just want some approximate values there, so that we can put them on a map at a very low zoom level and they kind of look right. If you load your GPS from INSPIRE you're not using it right...

Finally I agree. By having some records with super precise data and other with approximate one, we risk more to confuse our users. Plus it's a promise we are not going to maintain because we can't really keep up with maintain up-to-date addresses of everything. If we were to distinguish between approximate and exact result we would need to make this explicit to the user with some visual clue, and this might unnecessary clutter the UI, without sensible benefits.

So :+1: to send to the API just the city AND the zip code AND the country (zip code being necessary to distinguish city homonyms: I live in Thoiry in France, ... actually one of the 3 Thoiry in France...)
